internal/ingress/annotations/canary/main_test.go

```diff
@@ -0,0 +1,116 @@
+limitations under the License.
+*/
+
+package canary
+
+import (
+	api "k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/ingress-nginx/internal/ingress/annotations/parser"
+	"testing"
+
+	"k8s.io/ingress-nginx/internal/ingress/resolver"
+	"strconv"
+)
+
+func buildIngress() *extensions.Ingress {
+	defaultBackend := extensions.IngressBackend{
+		ServiceName: "default-backend",
+		ServicePort: intstr.FromInt(80),
+	}
+
+	return &extensions.Ingress{
+		ObjectMeta: metaV1.ObjectMeta{
+			Name:      "foo",
+			Namespace: api.NamespaceDefault,
+		},
+		Spec: extensions.IngressSpec{
+			Backend: &extensions.IngressBackend{
+				ServiceName: "default-backend",
+				ServicePort: intstr.FromInt(80),
+			},
+			Rules: []extensions.IngressRule{
+				{
+					Host: "foo.bar.com",
+					IngressRuleValue: extensions.IngressRuleValue{
+						HTTP: &extensions.HTTPIngressRuleValue{
+							Paths: []extensions.HTTPIngressPath{
+								{
+									Path:    "/foo",
+									Backend: defaultBackend,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func TestAnnotations(t *testing.T) {
+	ing := buildIngress()
+
+	data := map[string]string{}
+	ing.SetAnnotations(data)
+
+	tests := []struct {
+		title         string
+		canaryEnabled bool
+		canaryWeight  int
+		canaryHeader  string
+		canaryCookie  string
+		expErr        bool
+	}{
+		{"canary disabled and no weight", false, 0, "", "", false},
+		{"canary disabled and weight", false, 20, "", "", true},
+		{"canary disabled and header", false, 0, "X-Canary", "", true},
+		{"canary disabled and cookie", false, 0, "", "canary_enabled", true},
+		{"canary enabled and weight", true, 20, "", "", false},
+		{"canary enabled and no weight", true, 0, "", "", false},
+		{"canary enabled by header", true, 20, "X-Canary", "", false},
+		{"canary enabled by cookie", true, 20, "", "canary_enabled", false},
+	}
+
+	for _, test := range tests {
+		data[parser.GetAnnotationWithPrefix("canary")] = strconv.FormatBool(test.canaryEnabled)
+		data[parser.GetAnnotationWithPrefix("canary-weight")] = strconv.Itoa(test.canaryWeight)
+		data[parser.GetAnnotationWithPrefix("canary-by-header")] = test.canaryHeader
+		data[parser.GetAnnotationWithPrefix("canary-by-cookie")] = test.canaryCookie
+
+		i, err := NewParser(&resolver.Mock{}).Parse(ing)
+		if test.expErr {
+			if err == nil {
+				t.Errorf("%v: expected error but returned nil", test.title)
+			}
+
+			continue
+		}
+
+		u, ok := i.(*Config)
+		if !ok {
+			t.Errorf("%v: expected an External type", test.title)
+		}
+		if u.Enabled != test.canaryEnabled {
+			t.Errorf("%v: expected \"%v\", but \"%v\" was returned", test.title, test.canaryEnabled, u.Enabled)
+		}
+		if u.Weight != test.canaryWeight {
+			t.Errorf("%v: expected \"%v\", but \"%v\" was returned", test.title, test.canaryWeight, u.Weight)
+		}
```

Comment:
Is there a reason why there isn't a check for `canaryHeader` and `canaryCookie`?

Reply:
No, I just forgot to add those checks. I'll add them